Re-implement calc() to more closely align with current spec

weinig opened the pull request:
#### e6902e0483c51982df75abe81235860eae084e35
<pre>
Re-implement calc() to more closely align with current spec
<a href="https://bugs.webkit.org/show_bug.cgi?id=278244">https://bugs.webkit.org/show_bug.cgi?id=278244</a>

Reviewed by Antti Koivisto.

Re-implements CSS calc() to more closely align with current spec.

Structurally, the two main components replaced are platform level
`CalcExpressionNode` tree, replaced by `Calculation::Tree`, and the
CSS level `CSSCalcExpressionNode` tree, replaced by `CSSCalc::Tree`.

The new implementations remove the use of inheritance / virtual
dispatch to implement the hierarchies, instead using std::variant
based model, as well as the use of reference counting, instead
any allocated is done via UniqueRef. There was practically no
sharing in the old code path, so the reference counting was not
being utilized.

By removing the vtable and allowing for leaf nodes to be completely
inline we can save quite a bit of memory (though I don&apos;t think it
is likely that calc() is where memory is being used). For a simple
calc() like:

    calc(10% + 5em)

The old code would use:

    CSSCalcValue = 8 (CSSValue) + 8 (pointer to CSSCalcOperationNode [sum]) + 1 (range) + 7 (padding) = 40
    CSSCalcOperationNode [sum] = 32 (CSSCalcExpressionNode) + 8 (operator) + 16 (children vector) + 1 (range) + 1 (is root) + 6 (padding) = 64
    CSSCalcPrimitiveValueNode [%] = 32 (we can consider CSSPrimitiveValue to be free due to value pool)
    CSSCalcPrimitiveValueNode [em] = 32 (we can consider CSSPrimitiveValue to be free due to value pool)

    Total = 168 bytes (+ malloc overhead of four allocations)

where as the new implementation uses:

    CSSCalcValue = 8 (CSSValue) + 32 (CSSCalc::Tree) = 40
    Sum (pointed to by Tree) = 16 (children vector) + 48 (vector buffer with two Child nodes) = 64 bytes
    Percent (aka 10%) = free (inline in vector buffer)
    Dimension (aka 5em) = free (inline in vector buffer)

    Total = 104 bytes (+ malloc overhead of two allocations)

There is also plenty left to optimize if the want arises. Potential
profitable optimizations to try are:

   - Reducing the size of the Child std::variant from 24 bytes down
     to 16 bytes by either:
         a) replacing the use of std::variant with a custom type that
            allows packing the metadata of the leaf nodes with the tag.
         b) removing the need for for leaf node metadata by exploding
            it into individual structs (e.g. instead of one type called
            CanonicalDimension with a dimension value, 6 types called
            CanonicalDimensionLength, CanonicalDimensionAngle, ...).
   - Use an arena to allocate the non-leaf nodes and use 16bit offsets
     rather than pointers (we can constrain how many nodes are in calc()
     quite a bit).
   - Use a concise encoding for tree (RPN perhaps) and remove the need
     for pointers or offsets entirely.

To make working with the trees more ergonomic, each operator node has
been made &quot;tuple-like&quot;, meaning it can be destructured as well as used
by functions like WTF::apply(). The allows writing generic traversals
and constructions of operators, greatly reducing the amount of duplicate
code needed.

There are still a few failing calc() tests. The are mainly caused by
callers that don&apos;t pass conversion data when evaluating calc. They
will be resolved in followup changes.

* LayoutTests/TestExpectations:
    Un-skip a test that was asserting in the old calc() code but no longer does.
* LayoutTests/fast/css/calc-parsing-expected.txt:
* LayoutTests/fast/css/calc-parsing.html:
* LayoutTests/fast/css/object-position/parsing-object-position-expected.txt:
* LayoutTests/fast/css/object-position/parsing-object-position.html:
* LayoutTests/fast/shapes/shape-outside-floats/shape-outside-animation-expected.txt:
* LayoutTests/fast/shapes/shape-outside-floats/shape-outside-animation.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-color-mix-function-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/clip-path-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/register-property-syntax-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-functions/xywh-function-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/animation/translate-composition-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/cssScale.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/background-size-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/font-weight-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/acos-asin-atan-atan2-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/calc-in-media-queries-with-mixed-units-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/calc-nesting-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/getComputedStyle-calc-mixed-units-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/hypot-pow-sqrt-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/minmax-percentage-serialize-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/round-function-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/round-mod-rem-computed-expected.txt:
    Test changes.

* Source/WTF/wtf/StdLibExtras.h:
(WTF::apply_impl):
(WTF::apply):
    Adds WTF implementation of std::apply that works with tuple-like objects.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    Add new files.

* Source/WebCore/css/BasicShapeConversion.cpp:
* Source/WebCore/css/CSSCustomPropertyValue.cpp:
* Source/WebCore/css/CSSGradientValue.cpp:
* Source/WebCore/css/CSSHelper.h: Removed.
* Source/WebCore/css/CSSPrimitiveValue.cpp:
* Source/WebCore/css/CSSPrimitiveValue.h:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
* Source/WebCore/css/CSSToStyleMap.cpp:
* Source/WebCore/css/CSSUnits.cpp:
* Source/WebCore/css/CSSUnits.h:
* Source/WebCore/css/calc/CSSCalcCategoryMapping.cpp: Removed.
* Source/WebCore/css/calc/CSSCalcExpressionNode.cpp: Removed.
* Source/WebCore/css/calc/CSSCalcExpressionNode.h: Removed.
* Source/WebCore/css/calc/CSSCalcExpressionNodeParser.cpp: Removed.
* Source/WebCore/css/calc/CSSCalcExpressionNodeParser.h: Removed.
* Source/WebCore/css/calc/CSSCalcInvertNode.cpp: Removed.
* Source/WebCore/css/calc/CSSCalcInvertNode.h: Removed.
* Source/WebCore/css/calc/CSSCalcNegateNode.cpp: Removed.
* Source/WebCore/css/calc/CSSCalcNegateNode.h: Removed.
* Source/WebCore/css/calc/CSSCalcOperationNode.cpp: Removed.
* Source/WebCore/css/calc/CSSCalcOperationNode.h: Removed.
* Source/WebCore/css/calc/CSSCalcPrimitiveValueNode.cpp: Removed.
* Source/WebCore/css/calc/CSSCalcPrimitiveValueNode.h: Removed.
* Source/WebCore/css/calc/CSSCalcSymbolNode.cpp: Removed.
* Source/WebCore/css/calc/CSSCalcSymbolNode.h: Removed.
* Source/WebCore/css/calc/CSSCalcSymbolTable.h:
* Source/WebCore/css/calc/CSSCalcTree+CalculationValue.cpp: Added.
* Source/WebCore/css/calc/CSSCalcTree+CalculationValue.h: Added.
* Source/WebCore/css/calc/CSSCalcTree+ComputedStyleDependencies.cpp: Added.
* Source/WebCore/css/calc/CSSCalcTree+ComputedStyleDependencies.h: Added.
* Source/WebCore/css/calc/CSSCalcTree+Evaluation.cpp: Added.
* Source/WebCore/css/calc/CSSCalcTree+Evaluation.h: Added.
* Source/WebCore/css/calc/CSSCalcTree+NumericIdentity.h: Added.
* Source/WebCore/css/calc/CSSCalcTree+Parser.cpp: Added.
* Source/WebCore/css/calc/CSSCalcTree+Parser.h: Added.
* Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp: Added.
* Source/WebCore/css/calc/CSSCalcTree+Serialization.h: Added.
* Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp: Added.
* Source/WebCore/css/calc/CSSCalcTree+Simplification.h: Added.
* Source/WebCore/css/calc/CSSCalcTree+Traversal.h: Added.
* Source/WebCore/css/calc/CSSCalcTree.cpp: Added.
* Source/WebCore/css/calc/CSSCalcTree.h: Added.
* Source/WebCore/css/calc/CSSCalcType.cpp: Added.
* Source/WebCore/css/calc/CSSCalcType.h: Added.
* Source/WebCore/css/calc/CSSCalcValue.cpp:
* Source/WebCore/css/calc/CSSCalcValue.h:
* Source/WebCore/css/parser/CSSCalcParser.cpp:
* Source/WebCore/css/parser/CSSCalcParser.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Angle.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+CSSPrimitiveValueResolver.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+CSSPrimitiveValueResolver.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+IntegerDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Length.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Number.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Percent.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+RawResolver.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Resolution.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Time.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+UnevaluatedCalc.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+UnevaluatedCalc.h:
* Source/WebCore/css/typedom/CSSNumericValue.cpp:
* Source/WebCore/css/typedom/CSSNumericValue.h:
* Source/WebCore/css/typedom/CSSStyleValueFactory.cpp:
* Source/WebCore/css/typedom/CSSUnitValue.cpp:
* Source/WebCore/css/typedom/CSSUnitValue.h:
* Source/WebCore/css/typedom/StylePropertyMap.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathClamp.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathClamp.h:
* Source/WebCore/css/typedom/numeric/CSSMathInvert.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathInvert.h:
* Source/WebCore/css/typedom/numeric/CSSMathMax.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathMax.h:
* Source/WebCore/css/typedom/numeric/CSSMathMin.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathMin.h:
* Source/WebCore/css/typedom/numeric/CSSMathNegate.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathNegate.h:
* Source/WebCore/css/typedom/numeric/CSSMathProduct.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathProduct.h:
* Source/WebCore/css/typedom/numeric/CSSMathSum.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathSum.h:
* Source/WebCore/css/typedom/numeric/CSSMathValue.cpp: Added.
* Source/WebCore/css/typedom/numeric/CSSMathValue.h:
* Source/WebCore/css/typedom/transform/CSSScale.cpp:
* Source/WebCore/platform/Length.cpp:
* Source/WebCore/platform/calc/CalcExpressionBlendLength.cpp: Removed.
* Source/WebCore/platform/calc/CalcExpressionBlendLength.h: Removed.
* Source/WebCore/platform/calc/CalcExpressionInversion.cpp: Removed.
* Source/WebCore/platform/calc/CalcExpressionInversion.h: Removed.
* Source/WebCore/platform/calc/CalcExpressionLength.cpp: Removed.
* Source/WebCore/platform/calc/CalcExpressionLength.h: Removed.
* Source/WebCore/platform/calc/CalcExpressionNegation.cpp: Removed.
* Source/WebCore/platform/calc/CalcExpressionNegation.h: Removed.
* Source/WebCore/platform/calc/CalcExpressionNode.cpp: Removed.
* Source/WebCore/platform/calc/CalcExpressionNode.h: Removed.
* Source/WebCore/platform/calc/CalcExpressionNumber.cpp: Removed.
* Source/WebCore/platform/calc/CalcExpressionNumber.h: Removed.
* Source/WebCore/platform/calc/CalcExpressionOperation.cpp: Removed.
* Source/WebCore/platform/calc/CalcExpressionOperation.h: Removed.
* Source/WebCore/platform/calc/CalcOperator.cpp: Removed.
* Source/WebCore/platform/calc/CalcOperator.h: Removed.
* Source/WebCore/platform/calc/CalculationCategory.cpp:
* Source/WebCore/platform/calc/CalculationCategory.h:
* Source/WebCore/platform/calc/CalculationExecutor.h: Added.
* Source/WebCore/platform/calc/CalculationOperator.cpp: Added.
* Source/WebCore/platform/calc/CalculationOperator.h: Added.
* Source/WebCore/platform/calc/CalculationTree+Copy.cpp: Added.
* Source/WebCore/platform/calc/CalculationTree+Copy.h: Added.
* Source/WebCore/platform/calc/CalculationTree+Evaluation.cpp: Added.
* Source/WebCore/platform/calc/CalculationTree+Evaluation.h: Added.
* Source/WebCore/platform/calc/CalculationTree+Traversal.h: Added.
* Source/WebCore/platform/calc/CalculationTree.cpp: Added.
* Source/WebCore/platform/calc/CalculationTree.h: Added.
* Source/WebCore/platform/calc/CalculationValue.cpp:
* Source/WebCore/platform/calc/CalculationValue.h:
* Source/WebCore/platform/graphics/cg/GradientRendererCG.cpp:
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
* Source/WebCore/rendering/mathml/RenderMathMLBlock.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleResolveForFontRaw.cpp:
* Source/WebCore/svg/SVGLengthContext.cpp:
* Source/WebCore/svg/SVGSVGElement.cpp:

Canonical link: <a href="https://commits.webkit.org/282580@main">https://commits.webkit.org/282580@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/069e5dbcbb06829839d891720b0e59121696904e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/63568 "150 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42924 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/16165 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/67589 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/14176 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/65688 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50612 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/14456 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/51192 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9808 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66637 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39838 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/55048 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/31877 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36521 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/12422 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/13048 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/56680 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/58030 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12743 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/69285 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/62813 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/7515 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/12316 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/58499 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/7548 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/55134 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/58728 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14066 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/6270 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/84574 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/38745 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14904 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/39824 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/40936 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/39567 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
```diff
@@ -0,0 +1,1209 @@
+
+// MARK: Predicate: percentageResolveToDiminsion
+
+static bool percentageResolveToDiminsion(const SimplificationOptions& options)
```
Dimension

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-color-mix-function-expected.txt:
```diff
@@ -10,9 +10,9 @@ PASS e.style['color'] = "color-mix(in hsl, red 50%, blue)" should set the proper
 PASS e.style['color'] = "color-mix(in lch decreasing hue, red, hsl(120, 100%, 50%))" should set the property value
 FAIL e.style['color'] = "color-mix(in hsl, red calc(50% * sign(100em - 1px)), blue)" should set the property value Colors do not match.
-Actual:   color-mix(in hsl, red calc(50% * (100em - 1px)), blue)
+Actual:   color-mix(in hsl, red, blue)
 Expected: color-mix(in hsl, red calc(50% * sign(100em - 1px)), blue).
```
I don't think this is correct. We can't be sure of the sign until the font-size is resolved at computed value time.

e.g. `font-size: 0px;` will lead the sign function to be negative.

Specified value time serialization should always roundtrip, so sign() here should be preserved as-is.

weinig replied:
It’s not correct. The test is still failing.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/register-property-syntax-parsing-expected.txt:
```diff
@@ -188,7 +188,7 @@ PASS syntax:'<length>', initialValue:'var(--moo)' is invalid
 PASS syntax:'<length>', initialValue:'10%' is invalid
 PASS syntax:'<length>', initialValue:'calc(5px + 10%)' is invalid
-PASS syntax:'<length>', initialValue:'calc(5px * 3px / 6px)' is invalid
+FAIL syntax:'<length>', initialValue:'calc(5px * 3px / 6px)' is invalid assert_throws_dom: function "() => CSS.registerProperty({name: name, syntax: syntax, initialValue: initialValue, inherits: false})" did not throw
```
Do we have any history on why this was tested as invalid?

I do indeed see no reason for this to be invalid, but just want to make sure we've looked into why.

This test is also part of Interop 2024, so it's important to change if it's invalid.

weinig replied:
I don’t have history. I did bring it up in this issue though: https://github.com/w3c/csswg-drafts/issues/10752

nt1m replied:
Wait, I think this is indeed invalid, dividing `px` by `px` should cancel out the units, and give a number instead of a length ? The syntax here expects a `<length>`

weinig replied:
It's "5px * 3px / 6px". In unit algebra that is "length * length / length" which is equal to "length * number" which is equal to "length".

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/cssScale.tentative-expected.txt:
```diff
@@ -1,7 +1,7 @@
 PASS Constructing a CSSScale with an angle CSSUnitValue for the coordinates throws a TypeError
 PASS Constructing a CSSScale with a CSSMathValue that doesn't match <number> for the coordinates throws a TypeError
-PASS Constructing a CSSScale with an invalid division by px/px for the coordinates throws a TypeError
+FAIL Constructing a CSSScale with an invalid division by px/px for the coordinates throws a TypeError assert_throws_js: function "() => new CSSScale(coord, 0)" did not throw
```
Same question

weinig replied:
This one was incorrectly checked in by a chrome engineer. See https://github.com/w3c/csswg-drafts/issues/10733

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/background-size-expected.txt:
```diff
@@ -9,7 +9,7 @@ PASS Can set 'background-size' to a length: -3.14em
 PASS Can set 'background-size' to a length: calc(0px + 0em)
 PASS Can set 'background-size' to a percent: 0%
-PASS Can set 'background-size' to a percent: -3.14%
+FAIL Can set 'background-size' to a percent: -3.14% assert_approx_equals: expected -3.14 +/- 0.000001 but got 0
```
Same question here, what's the expected result here?

weinig replied:
I believe this test is wrong. background-size is defined to be "<bg-size> = [ <length-percentage [0,∞]> | auto ]{1,2} | cover | contain", so this would not roundtrip due to non-negative requirement.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/font-weight-expected.txt:
```diff
@@ -9,7 +9,7 @@ PASS Can set 'font-weight' to the 'bold' keyword: bold
 PASS Can set 'font-weight' to the 'lighter' keyword: lighter
 PASS Can set 'font-weight' to a number: 0
-FAIL Can set 'font-weight' to a number: -3.14 assert_approx_equals: expected -3.14 +/- 0.000001 but got -3
+FAIL Can set 'font-weight' to a number: -3.14 assert_approx_equals: expected -3.14 +/- 0.000001 but got 0
```
ditto

weinig replied:
Like above, font-weight is defined to [<number [1,1000]>](https://drafts.csswg.org/css-values-4/#number-value)], so the range check does not allow -3.

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-values/hypot-pow-sqrt-invalid-expected.txt:
```diff
@@ -35,7 +35,7 @@ PASS e.style['opacity'] = "pow(1, , 2)" should not set the property value
 PASS e.style['letter-spacing'] = "calc(1px * pow(2px, 3px))" should not set the property value
 PASS e.style['letter-spacing'] = "calc(sqrt(100px)" should not set the property value
-PASS e.style['letter-spacing'] = "hypot(2px, 40%)" should not set the property value
+FAIL e.style['letter-spacing'] = "hypot(2px, 40%)" should not set the property value assert_equals: expected "" but got "hypot(2px, 40%)"
```
I think we should change the WPT if it's incorrect. I do indeed see no reason for this to be disallowed.

weinig replied:
I'm pretty sure WebKit is wrong on this one, not the test. letter-spacing is supposed to `normal | <length>`, but we have it defined to `normal | <length-percentage>`. 

I am not sure why webkit has the definition the way it does, but it doesn't seem like the right change to make in this PR.

weinig replied:
Oh, I see, [css-text-4](https://drafts.csswg.org/css-text-4/#letter-spacing-property) actually changes it to <length-percentage>, so we are actually correct and the test should change.

nt1m commented on Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp:
```diff
@@ -0,0 +1,588 @@
+    case CSSUnitType::CSS_URI:
+    case CSSUnitType::CSS_VALUE_ID:
+    case CSSUnitType::CustomIdent:
+        break;
```
ASSERT_NOT_REACHED()

nt1m commented on Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp:
```diff
@@ -0,0 +1,588 @@
+    );
+}
+
+struct ChildRep {
```
nit: no abbreviations per webkit style

nt1m commented on Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp:
```diff
@@ -0,0 +1,588 @@
+    ValueRange range = ValueRange::All;
+};
+
+struct ParenSaver {
```
nit: no abbreviations

nt1m commented on Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp:
```diff
@@ -0,0 +1,588 @@
+    // 1. If the root of the calculation tree fn represents is a numeric value (number, percentage, or dimension), and the serialization being produced is of a computed value or later, then clamp the value to the range allowed for its context (if necessary), then serialize the value as normal and return the result.
+
+    if (state.stage == Stage::Computed) {
+        auto clampedFn = makeChildWithValueBasedOn(clampValue(fn.value, state.range), fn);
```
nit: no abbreviations

weinig replied:
`fn` is what the spec uses, so keeping it consistent was helpful for reading comprehension.

nt1m reviewed (commented):
I haven't taken a look at everything (not sure if I'll have the time to), but here are some initial comments.

anttijk commented on Source/WebCore/css/calc/CSSCalcTree.h:
```diff
@@ -0,0 +1,1153 @@
+    bool operator==(const IndirectNode<Op>& other) const { return type == other.type && op.get() == other.op.get(); }
+};
+
+using Node = std::variant<
```
nice tree representation

anttijk commented on Source/WebCore/css/calc/CSSCalcValue.cpp:
```diff
@@ -32,415 +33,268 @@
+double CSSCalcValue::doubleValueDeprecated(const CSSCalcSymbolTable& symbolTable) const
 {
-    return clampToPermittedRange(protectedExpressionNode()->doubleValue(primitiveType(), symbolTable));
+    if (m_tree.requiresConversionData)
+        ALWAYS_LOG_WITH_STREAM(stream << "ERROR: The value returned from CSSCalcValue::doubleValueDeprecated is likely incorrect as the calculation tree has unresolved units that require CSSToLengthConversionData to interpret. Update caller to use non-deprecated variant of this function.");
```
hope we can assert at some point

weinig replied:
Yep, that is the very next step for me!

anttijk commented on Source/WebCore/platform/calc/CalculationExecutor.h:
```diff
@@ -0,0 +1,481 @@
+namespace WebCore {
+namespace Calculation {
+
+// This file contains an implementation of math operations used by Calculation::Tree and CSSCalc::Tree.
```
It would be good to have an overview of the relationship between Calculation::Tree and CSSCalc::Tree somewhere. The former is used in style/Length values (via CalculationValue) and has everything resolved except percentages, right?

Wonder if there was some naming that would clarify things.

weinig replied:
Yeah, I thought a bunch about this and didn't come up with any names I loved.

PercentResolvingCalculation? 

I will add a nice overview in comment form for now, and continue working on a name.

anttijk replied:
or just PercentCalculation?

anttijk commented on Source/WebCore/css/calc/CSSCalcTree+Simplification.h:
```diff
@@ -0,0 +1,106 @@
+enum class Category : uint8_t;
+}
+
+namespace CSSCalc {
```
The file names are CSSCalcTree+Foo but the namespace is just CSSCalc. Is there some logic behind this I'm missing or could we make the naming match? Similarly with CalculationTree+Foo

weinig replied:
The idea was just about grouping:

- All the CSSCalc related stuff goes in the CSSCalc namespace.  All the Calculation related stuff goes in the Calculation namespace.
- All the CSSCalc "tree" related stuff, goes in files prefixed with CSSCalcTree. All the Calculation "tree" related stuff, goes in files prefixed with CalculationTree.

I was kinda basing the name spacing on things like WebCore::Style and WebCore::Layout, where we use the namespace as the prefix for the files (e.g. StyleBuilder.h has WebCore::Style::Builder, LayoutTreeBuilder.h has WebCore::Layout::TreeBuilder etc.).

This breaks down a bit when each header doesn't correspond with a type. So, for the CSSCalcTree and CalculationTree, there are a few header/cpp files that are just about some algorithm being used on the CSSCalc::Tree / Calculation::Tree, so in those cases, I thought the +Foo.h/cpp notation was the most clear..

anttijk reviewed (approved):
very cool

weinig committed:
Re-implement calc() to more closely align with current spec
https://bugs.webkit.org/show_bug.cgi?id=278244

Reviewed by Antti Koivisto.

Re-implements CSS calc() to more closely align with current spec.

Structurally, the two main components replaced are platform level
`CalcExpressionNode` tree, replaced by `Calculation::Tree`, and the
CSS level `CSSCalcExpressionNode` tree, replaced by `CSSCalc::Tree`.

The new implementations remove the use of inheritance / virtual
dispatch to implement the hierarchies, instead using std::variant
based model, as well as the use of reference counting, instead
any allocated is done via UniqueRef. There was practically no
sharing in the old code path, so the reference counting was not
being utilized.

By removing the vtable and allowing for leaf nodes to be completely
inline we can save quite a bit of memory (though I don't think it
is likely that calc() is where memory is being used). For a simple
calc() like:

    calc(10% + 5em)

The old code would use:

    CSSCalcValue = 8 (CSSValue) + 8 (pointer to CSSCalcOperationNode [sum]) + 1 (range) + 7 (padding) = 40
    CSSCalcOperationNode [sum] = 32 (CSSCalcExpressionNode) + 8 (operator) + 16 (children vector) + 1 (range) + 1 (is root) + 6 (padding) = 64
    CSSCalcPrimitiveValueNode [%] = 32 (we can consider CSSPrimitiveValue to be free due to value pool)
    CSSCalcPrimitiveValueNode [em] = 32 (we can consider CSSPrimitiveValue to be free due to value pool)

    Total = 168 bytes (+ malloc overhead of four allocations)

where as the new implementation uses:

    CSSCalcValue = 8 (CSSValue) + 32 (CSSCalc::Tree) = 40
    Sum (pointed to by Tree) = 16 (children vector) + 48 (vector buffer with two Child nodes) = 64 bytes
    Percent (aka 10%) = free (inline in vector buffer)
    Dimension (aka 5em) = free (inline in vector buffer)

    Total = 104 bytes (+ malloc overhead of two allocations)

There is also plenty left to optimize if the want arises. Potential
profitable optimizations to try are:

   - Reducing the size of the Child std::variant from 24 bytes down
     to 16 bytes by either:
         a) replacing the use of std::variant with a custom type that
            allows packing the metadata of the leaf nodes with the tag.
         b) removing the need for for leaf node metadata by exploding
            it into individual structs (e.g. instead of one type called
            CanonicalDimension with a dimension value, 6 types called
            CanonicalDimensionLength, CanonicalDimensionAngle, ...).
   - Use an arena to allocate the non-leaf nodes and use 16bit offsets
     rather than pointers (we can constrain how many nodes are in calc()
     quite a bit).
   - Use a concise encoding for tree (RPN perhaps) and remove the need
     for pointers or offsets entirely.

To make working with the trees more ergonomic, each operator node has
been made "tuple-like", meaning it can be destructured as well as used
by functions like WTF::apply(). The allows writing generic traversals
and constructions of operators, greatly reducing the amount of duplicate
code needed.

There are still a few failing calc() tests. The are mainly caused by
callers that don't pass conversion data when evaluating calc. They
will be resolved in followup changes.

* LayoutTests/TestExpectations:
    Un-skip a test that was asserting in the old calc() code but no longer does.
* LayoutTests/fast/css/calc-parsing-expected.txt:
* LayoutTests/fast/css/calc-parsing.html:
* LayoutTests/fast/css/object-position/parsing-object-position-expected.txt:
* LayoutTests/fast/css/object-position/parsing-object-position.html:
* LayoutTests/fast/shapes/shape-outside-floats/shape-outside-animation-expected.txt:
* LayoutTests/fast/shapes/shape-outside-floats/shape-outside-animation.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-color-mix-function-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-masking/parsing/clip-path-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-properties-values-api/register-property-syntax-parsing-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-shapes/shape-functions/xywh-function-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-transforms/animation/translate-composition-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/cssScale.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/background-size-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/font-weight-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/acos-asin-atan-atan2-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/calc-in-media-queries-with-mixed-units-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/calc-nesting-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/getComputedStyle-calc-mixed-units-003-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/hypot-pow-sqrt-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/minmax-percentage-serialize-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/round-function-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-values/round-mod-rem-computed-expected.txt:
    Test changes.

* Source/WTF/wtf/StdLibExtras.h:
(WTF::apply_impl):
(WTF::apply):
    Adds WTF implementation of std::apply that works with tuple-like objects.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    Add new files.

* Source/WebCore/css/BasicShapeConversion.cpp:
* Source/WebCore/css/CSSCustomPropertyValue.cpp:
* Source/WebCore/css/CSSGradientValue.cpp:
* Source/WebCore/css/CSSHelper.h: Removed.
* Source/WebCore/css/CSSPrimitiveValue.cpp:
* Source/WebCore/css/CSSPrimitiveValue.h:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
* Source/WebCore/css/CSSToStyleMap.cpp:
* Source/WebCore/css/CSSUnits.cpp:
* Source/WebCore/css/CSSUnits.h:
* Source/WebCore/css/calc/CSSCalcCategoryMapping.cpp: Removed.
* Source/WebCore/css/calc/CSSCalcExpressionNode.cpp: Removed.
* Source/WebCore/css/calc/CSSCalcExpressionNode.h: Removed.
* Source/WebCore/css/calc/CSSCalcExpressionNodeParser.cpp: Removed.
* Source/WebCore/css/calc/CSSCalcExpressionNodeParser.h: Removed.
* Source/WebCore/css/calc/CSSCalcInvertNode.cpp: Removed.
* Source/WebCore/css/calc/CSSCalcInvertNode.h: Removed.
* Source/WebCore/css/calc/CSSCalcNegateNode.cpp: Removed.
* Source/WebCore/css/calc/CSSCalcNegateNode.h: Removed.
* Source/WebCore/css/calc/CSSCalcOperationNode.cpp: Removed.
* Source/WebCore/css/calc/CSSCalcOperationNode.h: Removed.
* Source/WebCore/css/calc/CSSCalcPrimitiveValueNode.cpp: Removed.
* Source/WebCore/css/calc/CSSCalcPrimitiveValueNode.h: Removed.
* Source/WebCore/css/calc/CSSCalcSymbolNode.cpp: Removed.
* Source/WebCore/css/calc/CSSCalcSymbolNode.h: Removed.
* Source/WebCore/css/calc/CSSCalcSymbolTable.h:
* Source/WebCore/css/calc/CSSCalcTree+CalculationValue.cpp: Added.
* Source/WebCore/css/calc/CSSCalcTree+CalculationValue.h: Added.
* Source/WebCore/css/calc/CSSCalcTree+ComputedStyleDependencies.cpp: Added.
* Source/WebCore/css/calc/CSSCalcTree+ComputedStyleDependencies.h: Added.
* Source/WebCore/css/calc/CSSCalcTree+Evaluation.cpp: Added.
* Source/WebCore/css/calc/CSSCalcTree+Evaluation.h: Added.
* Source/WebCore/css/calc/CSSCalcTree+NumericIdentity.h: Added.
* Source/WebCore/css/calc/CSSCalcTree+Parser.cpp: Added.
* Source/WebCore/css/calc/CSSCalcTree+Parser.h: Added.
* Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp: Added.
* Source/WebCore/css/calc/CSSCalcTree+Serialization.h: Added.
* Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp: Added.
* Source/WebCore/css/calc/CSSCalcTree+Simplification.h: Added.
* Source/WebCore/css/calc/CSSCalcTree+Traversal.h: Added.
* Source/WebCore/css/calc/CSSCalcTree.cpp: Added.
* Source/WebCore/css/calc/CSSCalcTree.h: Added.
* Source/WebCore/css/calc/CSSCalcType.cpp: Added.
* Source/WebCore/css/calc/CSSCalcType.h: Added.
* Source/WebCore/css/calc/CSSCalcValue.cpp:
* Source/WebCore/css/calc/CSSCalcValue.h:
* Source/WebCore/css/parser/CSSCalcParser.cpp:
* Source/WebCore/css/parser/CSSCalcParser.h:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Angle.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+CSSPrimitiveValueResolver.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+CSSPrimitiveValueResolver.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+IntegerDefinitions.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Length.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Number.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Percent.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+RawResolver.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Resolution.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Time.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+UnevaluatedCalc.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+UnevaluatedCalc.h:
* Source/WebCore/css/typedom/CSSNumericValue.cpp:
* Source/WebCore/css/typedom/CSSNumericValue.h:
* Source/WebCore/css/typedom/CSSStyleValueFactory.cpp:
* Source/WebCore/css/typedom/CSSUnitValue.cpp:
* Source/WebCore/css/typedom/CSSUnitValue.h:
* Source/WebCore/css/typedom/StylePropertyMap.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathClamp.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathClamp.h:
* Source/WebCore/css/typedom/numeric/CSSMathInvert.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathInvert.h:
* Source/WebCore/css/typedom/numeric/CSSMathMax.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathMax.h:
* Source/WebCore/css/typedom/numeric/CSSMathMin.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathMin.h:
* Source/WebCore/css/typedom/numeric/CSSMathNegate.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathNegate.h:
* Source/WebCore/css/typedom/numeric/CSSMathProduct.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathProduct.h:
* Source/WebCore/css/typedom/numeric/CSSMathSum.cpp:
* Source/WebCore/css/typedom/numeric/CSSMathSum.h:
* Source/WebCore/css/typedom/numeric/CSSMathValue.cpp: Added.
* Source/WebCore/css/typedom/numeric/CSSMathValue.h:
* Source/WebCore/css/typedom/transform/CSSScale.cpp:
* Source/WebCore/platform/Length.cpp:
* Source/WebCore/platform/calc/CalcExpressionBlendLength.cpp: Removed.
* Source/WebCore/platform/calc/CalcExpressionBlendLength.h: Removed.
* Source/WebCore/platform/calc/CalcExpressionInversion.cpp: Removed.
* Source/WebCore/platform/calc/CalcExpressionInversion.h: Removed.
* Source/WebCore/platform/calc/CalcExpressionLength.cpp: Removed.
* Source/WebCore/platform/calc/CalcExpressionLength.h: Removed.
* Source/WebCore/platform/calc/CalcExpressionNegation.cpp: Removed.
* Source/WebCore/platform/calc/CalcExpressionNegation.h: Removed.
* Source/WebCore/platform/calc/CalcExpressionNode.cpp: Removed.
* Source/WebCore/platform/calc/CalcExpressionNode.h: Removed.
* Source/WebCore/platform/calc/CalcExpressionNumber.cpp: Removed.
* Source/WebCore/platform/calc/CalcExpressionNumber.h: Removed.
* Source/WebCore/platform/calc/CalcExpressionOperation.cpp: Removed.
* Source/WebCore/platform/calc/CalcExpressionOperation.h: Removed.
* Source/WebCore/platform/calc/CalcOperator.cpp: Removed.
* Source/WebCore/platform/calc/CalcOperator.h: Removed.
* Source/WebCore/platform/calc/CalculationCategory.cpp:
* Source/WebCore/platform/calc/CalculationCategory.h:
* Source/WebCore/platform/calc/CalculationExecutor.h: Added.
* Source/WebCore/platform/calc/CalculationOperator.cpp: Added.
* Source/WebCore/platform/calc/CalculationOperator.h: Added.
* Source/WebCore/platform/calc/CalculationTree+Copy.cpp: Added.
* Source/WebCore/platform/calc/CalculationTree+Copy.h: Added.
* Source/WebCore/platform/calc/CalculationTree+Evaluation.cpp: Added.
* Source/WebCore/platform/calc/CalculationTree+Evaluation.h: Added.
* Source/WebCore/platform/calc/CalculationTree+Traversal.h: Added.
* Source/WebCore/platform/calc/CalculationTree.cpp: Added.
* Source/WebCore/platform/calc/CalculationTree.h: Added.
* Source/WebCore/platform/calc/CalculationValue.cpp:
* Source/WebCore/platform/calc/CalculationValue.h:
* Source/WebCore/platform/graphics/cg/GradientRendererCG.cpp:
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
* Source/WebCore/rendering/mathml/RenderMathMLBlock.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleResolveForFontRaw.cpp:
* Source/WebCore/svg/SVGLengthContext.cpp:
* Source/WebCore/svg/SVGSVGElement.cpp:

Canonical link: https://commits.webkit.org/282580@main

rkirsling commented on Source/WebCore/css/calc/CSSCalcTree+NumericIdentity.h:
```diff
@@ -0,0 +1,303 @@
+    DPCM
+};
+
+constexpr uint8_t numberOfNumericIdentityTypes = std::to_underlying(NumericIdentity::DPCM) + 1;
```
This is a C++23 feature; short of implementing this in WTF, I think we should just use a normal `static_cast<uint8_t>`.

weinig replied:
Is it breaking anything? I didn't see any issues with it on EWS. We are building with c++23 already I believe (@achristensen07 please correct me if I am wrong).

achristensen07 replied:
We do have WTF::enumToUnderlyingType, but now that we are using C++23, we should be migrating it to std::to_underlying.  Does this cause playstation build failures?

nt1m replied:
Everything was changed to static_cast : https://github.com/WebKit/WebKit/commit/baaae40b9bf25c5b292c1c0bda83a74977c34715
